### DISCUSSION TITLE
Updating sites PnP Powershell

### DISCUSSION THREAD

**Discussion by pratheep-r** [pnp/powershell#791] [Q&A] [answered]
I am trying the below to turn off the offline availability on SharePoint Online sites,

` Cls

#Parameters
$SiteURL = "https://tenantname.sharepoint.com/"

$Cred = Get-Credential
 
#Connect to SharePoint Online site
Connect-PnPOnline -Url $SiteURL -Credentials $Cred

#Getting all the site collections
$SiteCollections = Get-PnPTenantSite -Filter "Url -like 'https://tenantname.sharepoint.com/sites/uat-dms'"

    #Loop through each site collection
    ForEach($Site in $SiteCollections)  
    {  
        Write-host -F Yellow "Disabling Offline Client availability" $Site.Url  
        Try {
            #Connect to site collection
            Connect-PnPOnline -Url $Site.Url -Credentials $Cred

            #Get the Web
            $Web = Get-PnPWeb -Includes ExcludeFromOfflineClient
            
            #Set "Allow items from this site to be downloaded to offline clients?" to "No"
            $web.ExcludeFromOfflineClient = $True
            $web.Update()
            Invoke-PnPQuery
            Write-Host "Sync Button is Disabled for the Site:" $Web.Url -f Green

        }

        Catch {
            write-host -f Red "`tError:" $_.Exception.Message
        }
    }```

I have 100,000 sites to loop through however it gets stuck nearly after running for an hour. I would appreciate any help or ideas please. 

Thanks.

**Comment by AndersRask** [comment, marked as answer]
First off, what version are you using, the old depricated SharePointPnPPowerShellOnline or PnP.PowerShell? The old one does not handle throttling, so it is a good idea to set up retry logic if you are being throttled or connections are dropped (429 and 5xx errors)

Secondly, please supply full exception (write out Get-PnPException). We cannot help you if you do not supply the specific exception you are receiving.

Regarding the code, you should use -ReturnConnection and assign a specific connection that you then remember to disconnect inside the loop in a Finally statement after catch:
```
try {
    $connection = Connect-PnPOnline -ReturnConnection -Url @Site.Url -Credentials $Cred
    ...
Finally {
    if ( $null -ne $connection ) {
        Disconnect-PnPOnline -Connection $connection
    }
}
```

**Reply by pratheep-r** [reply]
@AndersRask 
Thanks for the quick turn around. Seems -ReturnConnection helped. I also added 'Finally' block and Get-PnPException - thanks I have improved more. 

I had run against just over 7000 sites, which had taken 3 hours and 36 minutes. So to speak, PRD has more than 100,000 sites, which would take around 52 hours, if I have done my math correct. Would you suggest batching up to improve the performance and avoid throttling? If so could you help/suggest how to batch this scenario please?

**Reply by AndersRask** [reply]
That sounds a bit on the high side, but I guess it depends on latency.

You could batch this, but it would increase the risk of getting throttled. I usually just let it run its course...

If you decide to batch them, the easiest way is really just to start up several PowerShell prompts (start with a few and no more than 4 would be my recommendation). From each script just make sure you give them each a "stack" of sites. You could do this with a CAML query, or using Where-Object or something as simple as writing out all site URLS that you want to process into a CSV file ($SiteCollections | Export-Csv -Path sites.csv -Delimiter ";" -NoTypeInformation ) and then splitting it up into seperate CSV files that you feed to the script.

**Reply by pratheep-r** [reply]
Much appreciated.
